src/pages/index.js

```diff
@@ -5,86 +5,234 @@ import Section from "../components/Section.js";
+  handleAvatarSubmit
+);
+const imagePreviewModal = new PopupWithImage("#places-preview-modal");
+const deletePlaceModal = new PopupWithForm("#places-delete-modal", () => {});
 
+/* Event Listeners */
 editProfileModal.setEventListeners();
 imagePreviewModal.setEventListeners();
+updateAvatarModal.setEventListeners();
 addPlaceModal.setEventListeners();
-section.renderItems();
+
+/* Form Validation */
+deletePlaceModal.setEventListeners();
+avatarValidation.enableValidation();
 profileEditValidation.enableValidation();
 addPlaceValidation.enableValidation();
+
+/* DELETE Place Function */
+function handleDeleteClick(cardID, cardElement) {
+  deletePlaceModal.setSubmitHandler(() => {
+    deletePlaceModal.setLoading(true, "Removing...", "Yes");
+    api
+      .removePlace(cardID)
+      .then(() => {
+        cardElement.remove();
+        deletePlaceModal.close();
+      })
+      .catch((err) => {
+        console.error("Error deleting place:", err);
+      })
+      .finally(() => {
+        deletePlaceModal.setLoading(false, "Removing...", "Yes");
+      });
+  });
+  deletePlaceModal.open();
+}
+
+/* Create Cards */
+function getCardElement(cardData) {
+  const card = new Card(
+    cardData,
+    "#card-template",
+    handleImageClick,
+    handleDeleteClick,
+    handleLikeReact
+  );
+  return card.getView();
+}
+
+/* GET Cards */
+let section;
+api
+  .getInitialCards()
+  .then((items) => {
+    section = new Section(
+      {
+        items: items,
+        renderer: (cardData) => {
+          const cardElement = getCardElement(cardData);
+          return cardElement;
+        },
+      },
+      ".cards__list"
+    );
+    section.renderItems();
+  })
+  .catch((err) => console.error(err));
+
+/* GET Profile */
+const userInfo = new UserInfo();
+api
+  .getUserInfo()
+  .then((profileData) => {
+    userInfo.setUserInfo(profileData);
+  })
+  .catch((err) => console.error(err));
+
+/* PATCH Profile Edit Function */
+function handleProfileEditSubmit(profileInputValues) {
+  editProfileModal.setLoading(true);
+
+  const userData = {
+    name: profileInputValues.title,
+    about: profileInputValues.description,
+  };
+
+  api
+    .editUserInfo(userData.name, userData.about)
+    .then((updatedUserData) => {
+      userInfo.setUserInfo(updatedUserData);
+      editProfileModal.close();
+    })
+    .catch((err) => {
+      console.error("Error updating user info:", err);
+    })
+    .finally(() => {
+      editProfileModal.setLoading(false);
+    });
+}
+
+/* PATCH Profile Avatar Function */
+function handleAvatarSubmit({ url }) {
+  updateAvatarModal.setLoading(true);
+
+  api
+    .updateAvatar(url)
+    .then((res) => {
+      userInfo.updateAvatar(url);
+      updateAvatarModal.reset();
+      updateAvatarModal.close();
+      console.log("Success:", res);
+    })
+    .catch((err) => {
+      console.error("Error updating profile avatar:", err);
+    })
+    .finally(() => {
+      updateAvatarModal.setLoading(true, "Save");
```

Comment:
**Could be improved**


If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}

```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```